public/index.js

```diff
@@ -1,33 +1,149 @@
+    var postcode = noSpacesPostcode(postcodeInput.value);
+    if (regex.test(postcode)) {
+        return true
+    } else {
+        throw new Error('Please enter a valid postcode')
+    }
+}
 
+function categoryValid() {
+    var catValue = categoryInput.value.replace("&", "&amp;")
+    if (checkValues(fullCategoryList, catValue)) {
+        return true
+    } else {
+        throw new Error('Please enter a valid category')
+    }
 }
 
+postcodeInput.addEventListener("input", function() {
+    try {
+        postcodeValid()
+    } catch (err) {
+        displayError(err)
+    }
+})
+
+categoryInput.addEventListener("input", function() {
+    try {
+        categoryValid()
+    } catch (err) {
+        displayError(err)
+    }
+})
+
+form.addEventListener('input', function() {
+    try {
+        if (categoryValid() && postcodeValid()) {
+            button.disabled = false;
+        }
+    } catch (err) {
+        displayError(err)
+    }
+})
+
+
+/* ONCE SUBMIT ENABLED, CHECK POSTCODE AGAINST API AND MAKE REQUEST TO BACKEND */
+
+form.addEventListener('submit', function(event) {
+    event.preventDefault();
+    var postcode = noSpacesPostcode(postcodeInput.value)
+    var url = "http://api.postcodes.io/postcodes/" + postcode + "/validate"
+    XHRrequest(url, postcodeResponse);
+})
+
+function postcodeResponse(err, data) {
+    if (err) {
+        displayError(err)
+    } else {
+        if (data.result === false) {
+            displayError("Sorry, that postcode doesn't exist.")
+        } else if (data.result === true) {
+            var postcode = noSpacesPostcode(postcodeInput.value)
+            var catValue = categoryInput.value.replace("&", "&amp;");
+            var category = mapCategoryNametoID(fullCategoryList, catValue);
+            var url = "/results?postcode=" + postcode + "&category=" + category;
+            XHRrequest(url, updateMap)
+        }
+    }
+}
+
+/* MAP RELATED CODE */
+
+function initMap(lat, lng, zoom) {
+    var options = { center: { lat: lat, lng: lng }, zoom: zoom };
+    var map = new google.maps.Map(mapDiv, options);
+}
+// function updateMap(lat, lon, responseArr) {
+//     initMap(lat, lng);
+//     renderMarkers(responseArr);
+// }
+
+// function renderMarkers(responseArr) {
+//     // add markers to map
+// }
+
+
+/* XHR REQUEST */
+
 function XHRrequest(url, cb) {
     var xhr = new XMLHttpRequest();
     xhr.onreadystatechange = function() {
-      if (xhr.readyState === 4) {
-        if (xhr.status === 200) {
-          cb(null, xhr.responseText);
-        } else {
-          var errorMessage = xhr.responseText;
-          cb("Error " + url + " " + errorMessage);
+        if (xhr.readyState === 4) {
+            if (xhr.status === 200) {
+                cb(null, JSON.parse(xhr.responseText));
+            } else {
+                var errorMessage = xhr.responseText;
```

Comment:
Do we need to JSON.parse the error message? ie: 
var errorMessage = JSON(parse)xhr.responseText;

Reply:
Well spotted, will update now :)